text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+after the previous example). If errors are emitted at the point where the mutable reference is
+created, then there can be no such error here, as no mutable reference is ever created. `Cell::set`
+is a method that takes `&self`, not `&mut self`. Surely we can make interior mutability a special
+case, right? No; the only way to work around this is to make any reference to a type with interior
+mutability considered a mutation. Consequently, you could never have a reference to a type
+containing a `mut`-restricted, interior-mutable field. This is unacceptable, so interior mutability
+cannot be considered a mutation for the purposes of this restriction.
+
+[interior mutability]: https://doc.rust-lang.org/reference/interior-mutability.html
+
+### `struct` expressions are not allowed
+
+Given that the most common use for for `mut`-restricted fields is to ensure an invariant, it is
+important that the invariant be enforced. Consider the previous definition of `Time`. If you could
+write
+
+```rust
+Time {
+    hour: 32,
+    minute: 0,
+    second: 0,
+    nanosecond: 0,
+}
+```
+
+then the invariant would be violated, as there are only 24 hours in a day (numbered 0–23). Given
+that the invariant is not enforced by the type system, it cannot be enforced at all in this case. As
+a result, we have no choice but to disallow `struct` expressions for types with `mut`-restricted
+fields. This applies even when [functional update syntax][fru-syntax] is used, as invariants can
+rely on the value of other fields.
+
+[fru-syntax]: https://doc.rust-lang.org/stable/reference/expressions/struct-expr.html#functional-update-syntax
+
+Note that despite the name, `struct` expressions are not limited to `struct`s. They are used to
+initialize `enum` variants and `union`s as well. For `enum`s and `union`s, this restriction only
+applies to the specific variant being constructed. For example, the following is allowed:
+
+```rust
+pub enum Foo {
+    Alpha { mut(crate) x: u8 },
+    Beta { y: u8 },
+}
+
+// In another crate:
+Foo::Beta { y: 5 };
+```
+
+In this example, `Foo::Alpha { x: 5 }` is allowed when it is in the same crate as `Foo`. This is
+because `x` is not restricted within this scope, so the field can be freely mutated. Because of
+this, the previous concern about upholding variants is not applicable.
+
+## `mod` as a restriction scope
+
+`pub` can be restricted in a few ways. The most common is `pub(crate)`, which restricts visibility
+to the current crate. However, there are other ways to restrict visibility. You can also use
+`pub(super)` to restrict visibility to the parent module, and `pub(in path)` to restrict visibility
+to `path`, as long as that path is an ancestor of the location it is used. There is one additional
+case you have likely never encountered: `pub(self)`. The reason you have likely never seen this
+before is that it is completely useless. Why? `pub(self)` is identical to private, which is the
```

Comment:
```suggestion
before is that it is redundant: `pub(self)` is identical to private, which is the
```

It's redundant, but it isn't *completely* useless: I've seen uses for it in macros, for instance.